lib/Sema/TypeCheckCodeCompletion.cpp

```diff
@@ -1329,3 +1349,125 @@ void KeyPathTypeCheckCompletionCallback::sawSolution(
     Results.push_back({BaseType, /*OnRoot=*/(ComponentIndex == 0)});
   }
 }
+
+bool ArgumentTypeCheckCompletionCallback::Result::isSubscript() const {
+  return CallE && isa<SubscriptExpr>(CallE);
+}
+
+void ArgumentTypeCheckCompletionCallback::
+sawSolution(const constraints::Solution &S) {
+  GotCallback = true;
+
+  Type ExpectedTy = getTypeForCompletion(S, CompletionExpr);
+  if (!ExpectedTy)
+    return;
+
+  auto &CS = S.getConstraintSystem();
+
+  Expr *ParentCall = CompletionExpr;
+  while (ParentCall &&
+         !(isa<ApplyExpr>(ParentCall) || isa<SubscriptExpr>(ParentCall)))
+    ParentCall = CS.getParentExpr(ParentCall);
+
+  if (!ParentCall || ParentCall == CompletionExpr) {
+    assert(false && "no containing call?");
+    return;
+  }
+
+  auto *Locator = CS.getConstraintLocator(ParentCall);
+  auto *CalleeLocator = S.getCalleeLocator(Locator);
+  auto SelectedOverload = S.getOverloadChoiceIfAvailable(CalleeLocator);
+  if (!SelectedOverload)
+    return;
+
+  Type BaseTy = SelectedOverload->choice.getBaseType();
+  if (BaseTy)
+    BaseTy = S.simplifyType(BaseTy);
+
+  ValueDecl *FuncD = SelectedOverload->choice.getDeclOrNull();
+  Type FuncTy = S.simplifyType(SelectedOverload->openedType);
+
+  // For completion as the arg in a call to the implicit (keypath: _) method the
+  // solver can't know what kind of keypath is expected (e.g. KeyPath vs
+  // WritableKeyPath) so it ends up as a hole. Just assume KeyPath so we show
+  // the root type to users.
+  if (SelectedOverload->choice.getKind() == OverloadChoiceKind::KeyPathApplication) {
```

Comment:
I don’t quite understand when we need this special case. Do you have a test case at hand that fails without it? I haven’t had a chance to look at all of them yet.

Reply:
This was for an existing test case in IDE/complete_subscript.swift:
```
  let _ = MyStruct<Int>()[#^INSTANCE_INT_BRACKET^#
// INSTANCE_INT_BRACKET: Begin completions
// INSTANCE_INT_BRACKET-DAG: Decl[Subscript]/CurrNominal/Flair[ArgLabels]:        ['[']{#(x): Int#}, {#instance: Int#}[']'][#Int#];
// INSTANCE_INT_BRACKET-DAG: Pattern/CurrNominal/Flair[ArgLabels]:                ['[']{#keyPath: KeyPath<MyStruct<Int>, Value>#}[']'][#Value#];
// INSTANCE_INT_BRACKET: End completions
}
```
Without it the second result has no param or result type information (both are UnresolvedType), so it ends up being printed something like the below from memory:
```
['[']{#keyPath: _#}[']'][#_#];
```
I'll update the comment (it should say subscript rather than method too).

Reply:
OK, that makes sense. But wouldn’t the betters approach be that the constraint solver yields two solutions to us. IIUC, we implicitly have the following two subscripts defined on every type
```swift
subscript<Value>(keyPath keyPath: KeyPath<Self, Value>) -> Value { get }
subscript<Value>(keyPath keyPath: WritableKeyPath<Self, Value>) -> Value { get set }
```
so I would expect the constraint solver to return two solutions (one for `KeyPath` and one for `WritableKeyPath`).

Reply:
I suspect it's not actually implemented like that, since regular overloaded subscripts do give us two solutions. @xedin do you know how keypath application is set up in the solver at the moment?

Reply:
Unfortunately solver can't do anything about this. `subscript[keyPath: ...]` is not a regular overload so there is no parameter information for it, instead it infers whether it's `KeyPath` or `WritableKeyPath` based on the argument and in case of code completion it can't do that because there is no argument, so Nathan had to hardcode it to KeyPath to be safe if I remember correctly.

Reply:
Ah, OK. That’s unfortunate but probably the best solution for now.